_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+        } catch (e: Exception) {
+            connection.close() // (4)
+            throw SocketClientException("Error while exchanging data", e)
+        } finally {
+            connectionPool.releaseConnection(connection) // (5)
+        }
+    }
+
+    private fun exchange(connection: Connection, request: Request): Response { // (3)
+        val input = connection.getInputStream().bufferedReader()
+        val output = PrintWriter(connection.getOutputStream())
+
+        val requestRaw = request.toRaw()
+        output.print(requestRaw)
+        output.flush()
+
+        return Response.fromRaw(input)
+    }
+
+    override fun close() {
+        connectionPool.close()
+    }
+
+    fun stats() = SocketClientStats(connectionPool.stats())
+}
+
+open class SocketClientException(msg: String, throwable: Throwable? = null): RuntimeException(msg, throwable)
+data class SocketClientStats(val poolStats: ConnectionClientPoolStats)
+```
+
+We pass a bunch of arguments that were explained above. The `checkKeepAliveInterval`  (1) is how often the checker for cleaning up the old connections runs.  We pass it to the connection pool that we start on class initialization.
+
+In the `exchange()` function, we retrieve (lease or create) a connection from the pool (2). Then we do I/O converting data between strings and our structures (3). If we catch an exception, we have to close the connection (4). It probably means that there was a violation of the HTTP protocol, therefore we cannot continue sending more data on this connection. Whether we succeed or not, we should release the connection back to the pool (5).
+
+### ConnectionPool
+Here comes the difficult part. Firstly, I thought we could reuse the pool from the server, but the client’s pool is much more complicated. It supports connection leasing, limits connection per destination and checks connection inactivity in a different way.
+
+Let’s break it down one by one.
+
+#### Connection
+Just like in a server pool we need a wrapper over the socket that will track the activity on this socket.
+``` kotlin
+internal data class ConnectionDestination(val host: String, val port: Int) {
+    override fun toString(): String = "$host:$port"
+    fun toInetSocketAddress() = InetSocketAddress(host, port)
+}
+
+internal class Connection(
+        private val socket: Socket,
+        val destination: ConnectionDestination,
+        keepAliveTimeout: Duration,
+        checkKeepAliveInterval: Duration
+) {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(Connection::class.java)
+    }
+
+    /**
+     * Counter that is incremented on periodic inactivity check. It is reset on new connection activity
+     */
+    private var checks: Int = 0 // (1)
+
+    /**
+     * If the connection is checked for inactivity for inactiveChecksThreshold times and there was no activity
+     * it is assumed that connection is no longer active. It is optimization so we don't have to call time()
```

Comment:
It is *an* optimisation. 

http://www.vivquarry.com/wkshts/articles.html - "Articles are words which may or may not be used before a noun to demonstrate if the noun is new to the listener or reader"

http://www.grammar.cl/Notes/Articles.htm - "No Article. We use no article with: when we refer to general ideas"

Reply:
Damn articles........